Fatal Error: used more than once - [E0062]

# Fatal Error - `used more than once`
-  Fixing this error will improve gccrs to emit error codes similiar to rustc.
- You can view the same on [compiler-explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATKUcAZPAZMADl3ACNMYgkNUgAHVAVCOwYXNw9vOISkgQCg0JYIqK4Yq0wbZKECJmICVPdPH1LygUrqglyQ8MjoyyqauvTGvvbAzoLu4oBKS1QTYmR2DnNTGwBqADFUVFWAUgB2ACEdjQBBVfPV9VW8AGYfY5P9gBEHh6oGVZYmQIhJ3cOHhdVvQCJddjcnhstv8jqcgfCEZcQKsYoDEeirjFVgB6bGrSLEEjImiYWjoXYANg0qh2VNWCliZTwJPJLBImFWBAQhlWAgWaPR52eOxusMeeyeHGmtE4AFZeJ4OFpSKhOAAlMyghSzeYcnZeG48UgETRS6YAa28AA4AHR7ClWjReDQATgpLqtstl0hlHEkCtNKs4vAUIBiJqVUtIcFgMEQKFQLFidEi5EoaCTKaiwGQyGICgAtApmLEFAhUBYsAA3PALABqeEwAHcAPKMxVGmi0AiRUMQMKBsKBaoAT04RqHzGII5bYW0ZQjRozbEELYYtDHkdIWDCJmATjEtFD3F4WC%2BRnEW/wxAXeCrmGPyswqjKJh7494gR7vuVtDwYWIUcXCwQMCGIPAWA/aYqAMYAFAbZs20YD8ZEEEQxHYKRUPkJQ1EDXQuH0QxjE1fR/1DSBplQWJbAEY9eFQe9iHArAKKgZg2BATB8FohhSCrMQTEWZ0vB4SZpiaXiHAYZxXHqCRZV8GSOnyQo9HiRJeIGTwuEUjTsgYFSuiKSxSVvFphm0hTTOsXjWhqIzxhM8w2is3TejaRy1K4CSdQWCRpTlAMt1VDhLitCkCwpSRVgAcScJw1SEVYIHixLkq%2BcxIj%2BCBcEIEhdgNHzeAjLRxNIS1ZRiX1/VISDJBdG1PQpUSuApG4qU9N1SEVZVQpDMNjVNaYY3jWYCFiN80wgDNk3oYhglYRZTHMZBVi4G0HRtJ8eJIPB0D0fg0NEcQsKOnCVHULcCNIJtANiKD9CCnrA1Cls30m0FUCocLIuiuKEqSlK0qBzKe2IHKXEzebCsNSYSuGwK/V4SCqptKQbj2K0rQ6m4XSkDRpF6hjg0sQbSrNCrvEaq02tlLhMYpKQKT2SRJCtJ6OBuYK%2BtJimRrjUaICQWas2m0X5pAYgpBdGJq1rTAENbdsUK7cG%2BwHLdJ1HFDtenWd5xsFDl0YAg1w3QMdz3A9aCPFCz2Iy9lWvW970fXhn1fd8T3IQRSUDP8AKAjBFmVMCIMemCmDgpWkI7XhzvQ07pHOxRLvwjIDCMKXSMDtiqJo5J6JVJiWIfeAJLM5pPAgRw3KU9AvImTJNOSev9N4pvnKruzLLk9JCMkiphi7vQXP6fudI8hzRlUiZfLmfyfM5%2BUXpCzhVhWgg1o2hrtpSvKiGIWHiqGyNyoQTAmCwKJfip4pqs4WrUZiYmgw4Abw2G6NBaQcbPvFomOakRFqcQ4KoCKUUYog2SqlQGGUmBZQhh7XaLFDqyCTphFOsg054WuhkO6TAHoniRqvN%2Bb0Ppvl5D9LeO8bR7w0ClKGwDj76juPDM%2BZVpiX2vt0O%2BloH6c2fiAKqa9eYfzJl/c%2BFoQCSF3rpSQFJZReAdCzPYDNObczESTCR/MkZeB5johG0j%2BK9mSLIoAA%3D)

---
## I tried this code from [E0062](https://doc.rust-lang.org/error_codes/E0062.html):

```rust
struct Foo {
    x: i32,
}

fn main() {
    let x = Foo {
                x: 0,
                x: 0, // error: field `x` specified more than once
            };
}
```
---
## I expected to see this happen: 
- Give error like rustc 
```rust
error[E0062]: field `x` specified more than once
 --> <source>:8:17
  |
7 |                 x: 0,
  |                 ---- first use of `x`
8 |                 x: 0, // error: field `x` specified more than once
  |                 ^ used more than once

error: aborting due to previous error

For more information about this error, try `rustc --explain E0062`.
Compiler returned: 1
```
---
### Instead, this happened: 
- It gives fatal error
```rust
<source>:8:17: fatal error: used more than once
    8 |                 x: 0, // error: field `x` specified more than once
      |                 ^
compilation terminated.
Compiler returned: 1
```
---
### Meta

- What version of Rust GCC were you using, git sha 5406b633a1f5e6e3ae8da02589adf4ea003dec12.
- Godbolt version : `gccrs (Compiler-Explorer-Build-gcc-2f91d511200bf85558c9013b09a458c06edd1e02-binutils-2.40) 13.0.1 20230417 (experimental)`
---

--- a/gcc/rust/typecheck/rust-hir-type-check-struct-field.h
+++ b/gcc/rust/typecheck/rust-hir-type-check-struct-field.h
@@ -35,9 +35,9 @@ public:
 protected:
   void resolve (HIR::StructExprStructFields &struct_expr);
 
-  void visit (HIR::StructExprFieldIdentifierValue &field);
-  void visit (HIR::StructExprFieldIndexValue &field);
-  void visit (HIR::StructExprFieldIdentifier &field);
+  bool visit (HIR::StructExprFieldIdentifierValue &field);
+  bool visit (HIR::StructExprFieldIndexValue &field);
+  bool visit (HIR::StructExprFieldIdentifier &field);
 
 private:
   TypeCheckStructExpr (HIR::Expr *e);

--- a/gcc/rust/typecheck/rust-hir-type-check-struct.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-struct.cc
@@ -102,17 +102,24 @@ TypeCheckStructExpr::resolve (HIR::StructExprStructFields &struct_expr)
       switch (field->get_kind ())
 	{
 	case HIR::StructExprField::StructExprFieldKind::IDENTIFIER:
-	  visit (static_cast<HIR::StructExprFieldIdentifier &> (*field.get ()));
+	  ok = visit (
+	    static_cast<HIR::StructExprFieldIdentifier &> (*field.get ()));
 	  break;
 
 	case HIR::StructExprField::StructExprFieldKind::IDENTIFIER_VALUE:
-	  visit (
+	  ok = visit (
 	    static_cast<HIR::StructExprFieldIdentifierValue &> (*field.get ()));
 	  break;
 
 	case HIR::StructExprField::StructExprFieldKind::INDEX_VALUE:
-	  visit (static_cast<HIR::StructExprFieldIndexValue &> (*field.get ()));
+	  ok = visit (
+	    static_cast<HIR::StructExprFieldIndexValue &> (*field.get ()));
 	  break;
+	}
+
+      if (!ok)
+	{
+	  return;
 	}
 
       if (resolved_field_value_expr == nullptr)
@@ -238,16 +245,9 @@ TypeCheckStructExpr::resolve (HIR::StructExprStructFields &struct_expr)
   resolved = struct_def;
 }
 
-void
+bool
 TypeCheckStructExpr::visit (HIR::StructExprFieldIdentifierValue &field)
 {
-  auto it = fields_assigned.find (field.field_name.as_string ());
-  if (it != fields_assigned.end ())
-    {
-      rust_fatal_error (field.get_locus (), "used more than once");
-      return;
-    }
-
   size_t field_index;
   TyTy::StructFieldType *field_type;
   bool ok = variant->lookup_field (field.field_name.as_string (), &field_type,
@@ -255,7 +255,20 @@ TypeCheckStructExpr::visit (HIR::StructExprFieldIdentifierValue &field)
   if (!ok)
     {
       rust_error_at (field.get_locus (), "unknown field");
-      return;
+      return true;
+    }
+
+  auto it = adtFieldIndexToField.find (field_index);
+  if (it != adtFieldIndexToField.end ())
+    {
+      rich_location repeat_location (line_table, field.get_locus ());
+      auto prev_field_locus = it->second->get_locus ();
+      repeat_location.add_range (prev_field_locus);
+
+      rust_error_at (repeat_location, ErrorCode::E0062,
+		     "field %<%s%> specified more than once",
+		     field.field_name.as_string ().c_str ());
+      return false;
     }
 
   TyTy::BaseType *value = TypeCheckExpr::Resolve (field.get_value ().get ());
@@ -273,18 +286,14 @@ TypeCheckStructExpr::visit (HIR::StructExprFieldIdentifierValue &field)
       fields_assigned.insert (field.field_name.as_string ());
       adtFieldIndexToField[field_index] = &field;
     }
+
+  return true;
 }
 
-void
+bool
 TypeCheckStructExpr::visit (HIR::StructExprFieldIndexValue &field)
 {
   std::string field_name (std::to_string (field.get_tuple_index ()));
-  auto it = fields_assigned.find (field_name);
-  if (it != fields_assigned.end ())
-    {
-      rust_fatal_error (field.get_locus (), "used more than once");
-      return;
-    }
 
   size_t field_index;
   TyTy::StructFieldType *field_type;
@@ -292,7 +301,20 @@ TypeCheckStructExpr::visit (HIR::StructExprFieldIndexValue &field)
   if (!ok)
     {
       rust_error_at (field.get_locus (), "unknown field");
-      return;
+      return true;
+    }
+
+  auto it = adtFieldIndexToField.find (field_index);
+  if (it != adtFieldIndexToField.end ())
+    {
+      rich_location repeat_location (line_table, field.get_locus ());
+      auto prev_field_locus = it->second->get_locus ();
+      repeat_location.add_range (prev_field_locus);
+
+      rust_error_at (repeat_location, ErrorCode::E0062,
+		     "field %<%s%> specified more than once",
+		     field_name.c_str ());
+      return false;
     }
 
   TyTy::BaseType *value = TypeCheckExpr::Resolve (field.get_value ().get ());
@@ -310,18 +332,13 @@ TypeCheckStructExpr::visit (HIR::StructExprFieldIndexValue &field)
       fields_assigned.insert (field_name);
       adtFieldIndexToField[field_index] = &field;
     }
+
+  return true;
 }
 
-void
+bool
 TypeCheckStructExpr::visit (HIR::StructExprFieldIdentifier &field)
 {
-  auto it = fields_assigned.find (field.get_field_name ().as_string ());
-  if (it != fields_assigned.end ())
-    {
-      rust_fatal_error (field.get_locus (), "used more than once");
-      return;
-    }
-
   size_t field_index;
   TyTy::StructFieldType *field_type;
   bool ok = variant->lookup_field (field.get_field_name ().as_string (),
@@ -329,7 +346,20 @@ TypeCheckStructExpr::visit (HIR::StructExprFieldIdentifier &field)
   if (!ok)
     {
       rust_error_at (field.get_locus (), "unknown field");
-      return;
+      return true;
+    }
+
+  auto it = adtFieldIndexToField.find (field_index);
+  if (it != adtFieldIndexToField.end ())
+    {
+      rich_location repeat_location (line_table, field.get_locus ());
+      auto prev_field_locus = it->second->get_locus ();
+      repeat_location.add_range (prev_field_locus);
+
+      rust_error_at (repeat_location, ErrorCode::E0062,
+		     "field %<%s%> specified more than once",
+		     field.get_field_name ().as_string ().c_str ());
+      return false;
     }
 
   // we can make the field look like a path expr to take advantage of existing
@@ -358,6 +388,8 @@ TypeCheckStructExpr::visit (HIR::StructExprFieldIdentifier &field)
       fields_assigned.insert (field.get_field_name ().as_string ());
       adtFieldIndexToField[field_index] = &field;
     }
+
+  return true;
 }
 
 } // namespace Resolver

--- a/gcc/testsuite/rust/compile/repeated_constructor_fields.rs
+++ b/gcc/testsuite/rust/compile/repeated_constructor_fields.rs
@@ -1,0 +1,10 @@
+struct Foo {
+    x: i32,
+}
+
+fn main() {
+    let x = Foo {
+        x: 0,
+        x: 0, // { dg-error "field 'x' specified more than once" }
+    };
+}

